optional.h

```diff
@@ -0,0 +1,31 @@
+#include <cassert>
+#include <iostream>
+#include <string>
+
+template <class T>
+class Optional {
```

Comment:
Please, test `Optional` with the following class:
```
struct my_class {
    my_class(int) {}
};
```

Reply:
It doesn't work since my_class is not std::is_default_constructible, I guess I should add the requirement to the definition of Optional.

Reply:
I can also implement my optional with variant, in this case it does not need to have T as default_constructible. But  I am not sure it is fair:)

Reply:
And I could also allocate memory on the stack for T, keep inside a pointer of T and when I need to create an object of T I will use this memory. Thus also  I won't need a default_constructible type.

Reply:
If the optional requires a default-constructible T, why couldn't we just use more simple and less obfuscated std::pair<bool, T>?
Of course it should not have such a constraint.
Besides, the assignment requires not to use free store ("heap").

Reply:
But using pair<bool, T> was exactly my naive implementation (just didn't use explicitly the pair)!
I understand that we want it work also with non default-constructible and therefore my solution is not good enough.
I didn't suggest to allocate memory on the heap, because the ex explicitly mentioned it.
I suggest to allocate memory on the stack (char buf[sizeof(T)])  and when we actually use the T call the new expression with this buffer
T* tptr = new(buf) T(t).